### PR TITLE
IDC: build and display the new banner

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
- build the new RNA IDC banner
- display the banner instead of the existing one
- remove the old banner's code
- hide the Connection UI if connection is in IDC mode

#### Jetpack product discussion
p9dueE-3Mn-p2

#### Does this pull request change what data or activity we track or use?
New tracking events were not introduced, only migrated existing ones to the new IDC package.

##### Initial tests
1. Connect Jetpack, go to Jetpack Debug and activate the IDC Simulator.
2. Go to "Jetpack Debug -> IDC Simulator", enable "IDC Simulation" and enter a "Spoof URL".
3. Edit a post and save to make the site sync. Wait a few seconds.
4. Go to Jetpack Dashboard, confirm that the RNA IDC screen appears, and Jetpack Dashboard does not.
5. Open the browser's dev console, enter `localStorage.debug = "dops:analytics";` to see the tracking events. Keep the console open, and activate the "Persist Log" feature.
6. Go to a few non-Jetpack WP-admin pages, confirm that the RNA IDC banner appears above the main content.
7. The console should show two events being tracked: `jetpack_idc_notice_view` and `"notice-view" in group "jetpack-idc"`

##### Stay in safe mode

8. Click "Stay in Safe Mode". The page should reload, the IDC banners should disappear.
9. The console should log two events: `jetpack_idc_confirm_safe_mode` and `confirm-safe-mode`.
10. "Jetpack Safe Mode" button should appear in the top right corner, click it. The IDC banner should appear again.

##### Move your settings
11. Click "Move your settings", the "migrated successfully" screen should appear, click "Got it". The page should reload, IDC banners should disappear.
12. The console should log two events: `jetpack_idc_migrate` and `"migrate" in group "jetpack-idc"`.

##### Fresh connection
13. Go to the "IDC Simulator", disable simulation. Then disconnect Jetpack and connect again.
14. Enable the simulation, edit a post. You might need to wait a bit.
15. Go to Jetpack Dashboard, confirm that you see the IDC screen, and do not see the dashboard itself.
16. Keep the IDC screen browser tab open, in another tab go to the IDC simulator and disable the simulation. With simulation enabled you won't be able to connect the site.
17. Switch back to the IDC screen tab, click "Create a fresh connection". Confirm that you see the two events tracked in the console: `jetpack_idc_start_fresh` and `"start-fresh" in group "jetpack-idc"`.
18. Go through the connection process, confirm that you're redirected back to the Connection Manager page. IDC banners should disappear.

### CODE REVIEW

File: projects/packages/identity-crisis/src/class-ui.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+				'../build/index.js',
+				__FILE__,
+				array(
+					'in_footer'  => true,
+					'textdomain' => 'jetpack',
+				)
+			);
+			Assets::enqueue_script( 'jp_identity_crisis_banner' );
+			wp_add_inline_script( 'jp_identity_crisis_banner', static::get_initial_state(), 'before' );
+			wp_set_script_translations( 'jp_identity_crisis_banner', 'jetpack' );
```
Comment (anomiex) [inline projects/packages/identity-crisis/src/class-ui.php:65]: This should be unneeded, specifying the textdomain at line 60 should be having `Assets::register_script()` handle it.
```suggestion
```

**Review by anomiex — COMMENTED** [review]
I only really looked at the changes to the Webpack and other build stuff. One small comment inline. Overall I like the dropping of gulp, from a build system perspective it seems a lot cleaner now. How was the change from the JS side of things?

**Comment by sergeymitr** [comment]
Thanks @anomiex!

> How was the change from the JS side of things?

The only difference is that we still don't have Webpack building SCSS files directly, it's done through loading those files into JS files.
But we no longer have any separate styles in the package, so that's ok. We still have to solve this problem for the JITM package, but I didn't get to it yet.

**Comment by anomiex** [comment]
> The only difference is that we still don't have Webpack building SCSS files directly, it's done through loading those files into JS files. [...] We still have to solve this problem for the JITM package, but I didn't get to it yet.

Webpack can build the scss files directly if you list them as the files for the entry point rather than a JS file, but it'll generate a JS file for the bundle even after mini-css-extract-plugin extracts all the built CSS. We could use something like https://github.com/Automattic/remove-asset-webpack-plugin or https://github.com/webdiscus/webpack-remove-empty-scripts to suppress that JS file.

File: projects/packages/identity-crisis/src/class-ui.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		do_action( 'jetpack_identity_crisis_ui_init' );
+
+		$idc_data = Identity_Crisis::check_identity_crisis();
+
+		// TODO: replace the `jetpack_disconnect` check with a non-admin IDC screen.
+		if ( false === $idc_data || ! current_user_can( 'jetpack_disconnect' ) ) {
+			return;
+		}
+
+		$idc_data = Identity_Crisis::check_identity_crisis();
```
Comment (kbrown9) [inline projects/packages/identity-crisis/src/class-ui.php:42]: This looks like some leftover code (see line 35).

**Review by samiff — CHANGES_REQUESTED** [review]
> Open the browser's dev console, enter localStorage.debug = "dops:analytics"; to see the tracking events.

I haven't played much with the tracking stuff before, I imagine I was supposed to be seeing the different events logged to my browser console? Wasn't able to get this to work, even tried `localStorage.debug = "*";` with no luck. I was able to see the correct t.gif events in the Network tab however. Any idea on why the debug wasn't working or where I can read more about how that is setup?

> Go through the connection process, confirm that you're redirected back to the Connection Manager page. IDC banners should disappear.

On reconnecting I ended up on Calypso plans page. Otherwise the other testing instructions worked well for me 👍  And looks like the webpack built CSS is working nicely.

**Comment by samiff** [comment]
Ah, missed these two notices that came up in my error log:

```
[29-Nov-2021 21:01:29 UTC] PHP Notice:  Trying to access array offset on value of type bool in /usr/local/src/jetpack-monorepo/projects/packages/identity-crisis/src/class-ui.php on line 100
[29-Nov-2021 21:01:29 UTC] PHP Notice:  Trying to access array offset on value of type bool in /usr/local/src/jetpack-monorepo/projects/packages/identity-crisis/src/class-ui.php on line 101
```

```php
'wpcomHomeUrl'        => $idc_urls['wpcom_url'],
'currentUrl'          => $idc_urls['current_url'],
```

Probably from flipping switches in the Debug Tools, not sure if it'd come up naturally 🤷

**Review by kbrown9 — COMMENTED** [review]
I haven't tested yet so I'm not ready to approve this. I just have a few very minor comments about the code.

File: projects/packages/identity-crisis/src/class-ui.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		/**
+		 * Action called after initializing Connection UI Admin resources.
+		 *
+		 * @since 1.1.0
+		 */
+		do_action( 'jetpack_identity_crisis_ui_init' );
+
+		$idc_data = Identity_Crisis::check_identity_crisis();
+
+		// TODO: replace the `jetpack_disconnect` check with a non-admin IDC screen.
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply projects/packages/identity-crisis/src/class-ui.php:37]: Good catch.
We're essentially removing the non-admin IDC screen with this PR, as it's not implemented yet.
However, we just discussed it with @kbrown9, and decided that it's safer to put off this PR until we implement that non-admin screen.

**Comment by samiff** [comment]
With #22060 merged, could this be closed now?

**Comment by sergeymitr** [comment]
Good point @samiff, closing it now.

**Pull request closed without merging**